remotes/fixup.go

```diff
@@ -212,21 +247,109 @@ func fixupImage(ctx context.Context, baseImage bundle.BaseImage, cfg fixupConfig
 	return baseImage, nil
 }
 
-// fixupPlatform resolve a single image manifest out of a manifest list if a platform filter has been specified
-// it modifies the baseImage and fixupInfo accordingly
-func fixupPlatform(ctx context.Context, cfg fixupConfig, baseImage *bundle.BaseImage, fixupInfo *imageFixupInfo, sourceFetcher remotes.Fetcher) error {
-	if cfg.platform == "" ||
+func fixupPlatforms(ctx context.Context, cfg fixupConfig, baseImage *bundle.BaseImage, fixupInfo *imageFixupInfo, sourceFetcher sourceFetcherAdder, filter []platforms.Matcher) error {
+	if len(filter) == 0 ||
 		(fixupInfo.resolvedDescriptor.MediaType != ocischemav1.MediaTypeImageIndex && fixupInfo.resolvedDescriptor.MediaType != images.MediaTypeDockerSchema2ManifestList) {
 		// no platform filter if platform is empty, or if the descriptor is not an OCI Index / Docker Manifest list
 		return nil
 	}
+	if len(filter) == 1 {
+		return fixupSinglePlatform(ctx, cfg, baseImage, fixupInfo, sourceFetcher, filter[0])
+	}
 
-	plat, err := platforms.Parse(cfg.platform)
+	reader, err := sourceFetcher.Fetch(ctx, fixupInfo.resolvedDescriptor)
 	if err != nil {
 		return err
 	}
-	matcher := platforms.NewMatcher(plat)
+	defer reader.Close()
 
+	manifestBytes, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return err
+	}
+	var manifestList typelessManifestList
+	if err := json.Unmarshal(manifestBytes, &manifestList); err != nil {
+		return err
+	}
+	valid := 0
+	for _, d := range manifestList.Manifests {
+		if matchAny(d.Platform, filter) {
+			manifestList.Manifests[valid] = d
+			valid++
+		}
+	}
+	manifestList.Manifests = manifestList.Manifests[:valid]
+	manifestBytes, err = json.Marshal(&manifestList)
+	if err != nil {
+		return err
+	}
+	d := sourceFetcher.Add(manifestBytes)
+	descriptor := fixupInfo.resolvedDescriptor
+	descriptor.Digest = d
+	fixupInfo.resolvedDescriptor = descriptor
+	newRef, err := reference.WithDigest(fixupInfo.targetRepo, d)
+	if err != nil {
+		return err
+	}
+	baseImage.Image = newRef.String()
+	return nil
+}
+
+func matchAny(plat *ocischemav1.Platform, filter []platforms.Matcher) bool {
+	if plat == nil {
+		return false
+	}
+	for _, m := range filter {
+		if m.Match(*plat) {
+			return true
+		}
+	}
+	return false
+}
+
+type typelessManifestList struct {
+	Manifests []typelessDescriptor   `json:"manifests"`
+	Extras    map[string]interface{} `json:,inline`
+}
```

Comment:
Hopefully this round trips nicely and all. Not sure if `json.RawMessage` is preferable to `interface{}`, likely makes no difference I suspect.